### ISSUE TITLE
Why Mix CocoaPods and npm ?

### ISSUE THREAD

**Issue by cancan101** [auth0/react-native-lock-ios#2]
I don't understand the need to use CocoaPods at all.

There are plenty of other RN projects that get by with just npm:
* https://github.com/aksonov/react-native-tableview
* https://github.com/lwansbrough/react-native-camera (supports IOS and Android)
* https://github.com/umhan35/react-native-search-bar

They just have the user add the `xcodeproj` to the libraries in XCode.

TBH I don't think I have seen any react-native component that uses both CocoaPods and npm.

**Comment by hzalaz** [comment]
Because you are still building a native iOS library that has other native dependencies and if you are not using CocoaPods you'll have to do copy the xcodeproj of every dependency of Lock and Lock itself. Also we include not only code but resources like images and the UI (xib files) so you'll need to make sure those end up in your application bundle (and properly compiled). 
This process is highly error prone and will cause a lot of support request to us so we decided to rely only in CocoaPods for the time being.
There are some components that support both CP & npm e.g. Facebook RN SDK https://github.com/facebook/react-native-fbsdk and there is even an OSS effort to build a tool that handles both for you https://github.com/exponentjs/rem.

If you come up with a better way of doing this please feel free to submit a Pull Request and we'll review it and merge it.

**Comment by cancan101** [comment]
Isn't it possible to include a `.xcodeproj` folder in the npm package allowing a user that doesn't want to use pods to add the project to an existing project?

**Comment by hzalaz** [comment]
Maybe I wasn't clear enough before here 
> This process is highly error prone and will cause a lot of support request to us so we decided to rely only in CocoaPods for the time being.

but let me rephrase, it is possible to include it but is my choice to avoid that until we find a way to minimise the possible things that can go wrong (duplicate symbols, libraries linked twice incrementing the size of the app binary, etc) and also make it easy to include Lock without the need to write lengthy and complicated instructions. 
It's not as easy as `including the .xcodeproj file of the library in npm` as I explained in my previous comment, You need to take care that all the libs that react-native-lock-ios are properly compiled, linked along with it's resources (e.g. images) and CocoaPods does that for you without dealing with Xcode. 
If you don't want to use CocoaPods (for whatever reasons you have), I am sorry but this is the only guaranteed way that this library will work (and that we provide support if needed) for the time being until we can come up for a solution without CP.

**Comment by cancan101** [comment]
Also the instructions are not clear. This is really weird:
>Another necessary step you need to do is remove the React, RCTImage, etc. subprojects from your app's Xcode project.

and I would prefer NOT to deviate from how react-native is installed by the CLI. What is going to happen when I need to upgrade? Ideally I can just use: https://facebook.github.io/react-native/docs/upgrading.html#2-upgrade-your-project-templates.

**Comment by hzalaz** [comment]
They are all the sub projects inside your Xcode app project. You can see them inside the folder named  Library. The reason to do that is to avoid adding twice react library twice, one included by react cli and another by CocoaPods which will cause this issue #2. 
I have a change to clarify that step, but didn't have time today to upload it, so it's clear enough of you don't know how a RN project is built. 
 
About th update using the CLI, it should work since it will prompt you when it detects a diff, but I'll try to give it a try since RN sometimes ships breaking changes to the templates.

**Comment by cancan101** [comment]
Why not just use the react shipped  with cli and not add another with pod?

**Comment by hzalaz** [comment]
It will use the one shipped with cli but needs to be added using CocoaPods not using sub projects.

**Comment by cancan101** [comment]
I think you might need a..  In the path for node modules. See https://github.com/facebook/react-native-fbsdk/blob/master/README.md#option-using-cocoapods but I know for me the pod file is I  iOS subdir

**Comment by hzalaz** [comment]
Yes, that's part of the doc update I need to upload to the site since I forgot to use the template structure for the guide

**Comment by hzalaz** [comment]
After checking those two libs, and they don't have support for CocoaPods. So it's either not using Lock and wait for us to make it work without CP or don't use those libraries. 
There is also the alternative of adding the CocoaPods support for those libraries, it's just adding a file declaring what files the library has, but that requires a PR to those libraries repositories.

**Comment by cancan101** [comment]
Right now I require those libraries. I personally don't know enough about CP to confidently submit PRs, but even so, I foresee there being other RN libraries that I use down the road that don't currently ship with CP support.

**Comment by cancan101** [comment]
There are actually three libs used (same three in the OP).

**Comment by hzalaz** [comment]
Yes, missed that one when reading the linker message.

**Comment by hzalaz** [comment]
About the PR for CP on there libs, I see some of them without that support and other that also include the podspec file. Not sure why FB does not standardise how a lib should be built, this project https://github.com/exponentjs/rem started as a clean way to handle this but seems that FB is not contributing to it. For this kind of small libs it's easy to build and maintain a `.xcodeproj` file but if you want to reuse code of libs other than RN or iOS SDK you are technically screwed and need to resort to some bad practices like distributing binary files since NPM can't handle that. 
The reason I dislike this approach (and most community consider it a bad practice), is that they can easily be broken by a new release of Xcode, specially if they are coded in Swift since its not [ABI compatible](https://en.wikipedia.org/wiki/Application_binary_interface) between versions. (ObjC is ABI compatible but another issues might arise).
Also there is a security concern since you don't know what code is bundled there so you need to make sure you are not including some sort of malware, you can double check that the binary comes from a trusted source but you need to be *always* vigilant. (You can disassembly the binary but that is something most RN devs wouldn't dare to do).

To sum up, [Lock for iOS](https://github.com/auth0/Lock.iOS-OSX) has already a xcodeproj that we can use but it relies on a different dependency manager (Carthage) to fetch it's dependencies so that's doesn't help us, so we might end up providing a static library of Lock for iOS inside react-native-lock-ios so you can use an xcodeproj (a static lib and a couple bundle files with assets and UI components).

I don't have an ETA on this since we are working on other things right now.

**Comment by secobarbital** [comment]
I had to link ```libReact.a``` when I hit this error.

**Comment by cancan101** [comment]
I added `libReact.a` and the build now completed. I can `import Auth0Lock from 'react-native-lock-ios';` however when I call  `lock.show`, I get:
```
Uncaught TypeError: Cannot read property 'init' of undefined(…)
show @ auth0-lock.js:18
```

It appears that `LockModule` is not defined.

**Comment by hzalaz** [comment]
@cancan101 not sure what is wrong, the error is because this require `var LockModule = require('react-native').NativeModules.Auth0LockModule;` yields `undefined` but its working for me (check the sample in this repo). Can you send a project that can reproduce this issue to support@auth0.com?

**Comment by cancan101** [comment]
sent.

**Comment by cancan101** [comment]
In asking about packaging for this type of project, the consensus on the [reactiflux channel](http://www.reactiflux.com/) was to the more manual "adding the needed files" and waiting for [REM](https://github.com/exponentjs/rem) rather than using CP.

**Comment by hzalaz** [comment]
I knew that by reading other email list and GH issues (never had the chance to join that channel but i'll keep it in mind), but that works for tiny libs with no dependencies at all and are only used for react-native which is not the case for Lock for iOS which should work without RN. 
Haven't checked on android yet but I am mostly sure it needs to rely on gradle at least to fetch the dependencies of the java libs (I really hope I don't have to manually handle the classpath which will be a nightmare) so IMO this is the same. I have a plan to strip most of Lock for iOS dependencies but if you need native integrations e.g. for logins with FB you're back to square one.

I am closing this issue since there is not much to add. I'll create a separate one to add the "manual" import with `.xcodeproj` file and will answer your undefined issue via email

**Comment by hzalaz** [comment]
For reference the issue is #4

**Comment by cancan101** [comment]
@hzalaz And what about the upgrade question (making that this still works: https://facebook.github.io/react-native/docs/upgrading.html#2-upgrade-your-project-templates):

**Comment by hzalaz** [comment]
It will complain about your app's `.xcodeproj` since CocoaPods modified it (and you by removing the libraries subproject), for the rest it will work. In that case you have 2 alternatives, tell it to leave the file as it is which should be safe unless RN has a big change on how it builds in iOS (which should not happen a lot) or overwrite the `.xcodeproj` and apply the CocoaPods changes again (remove libraries subproject and run `pod install` again).

**Comment by hzalaz** [comment]
Also It might complain if you changed the `AppDelegate.m` file which is pretty common if you need to handle stuff like Push Notifications, Handoff or using FB SDK which requires some changes in that file.

**Comment by hzalaz** [comment]
My suggestion will be to apply those changes yourself since these kind of tools tend to break things easily and it's good to know what's under the hood, specially with this kind of fwk.
